### ISSUE TITLE
Very slow data retrival using @DocumentReference or @DBRef annotation

### ISSUE THREAD

**Issue by dominikpiegdon** [spring-projects/spring-data-mongodb#4417]
Hello,

I am writing to discuss an issue I encountered while fetching data from a MongoDB database using the MongoRepository. The data models involved in this scenario are as follows:

```
@Document(collection = "calendar_event")
class CalendarEvent(
    @Id
    private String id,

    private OffsetDateTime start,

    private OffsetDateTime end,

    private String description,

    private CalendarEventType type,

    @DBRef
    private List<User> participants
) implements Serializable {
    // Class implementation
}
```

```
@Document(collection = "user")
class User(
    @Id
    private String id,

    private String email,

    private String password,

    private String firstName,

    private String lastName
) {
    // Class implementation
}
```

Although the insertion of these objects into the database was relatively quick, the read operation (calendarEventRepository.findAll()) is experiencing severe slowness. After enabling detailed logging at the TRACE level in MongoDB, I discovered that the findAll operation executes a separate query for each CalendarEvent document to retrieve the associated users. I am curious whether this behavior aligns with the expected behavior of the framework. 

I actually debug the spring-data-mongodb code and I found out that method readReference of ReferenceLookupDelegate object is executed for each property of each entity. It means that if there is @DocumentReference or @DBRef on the property, for each property a new query will be run if I am no mistaken.

It seems that the only way of performing single query fetch is to create an Aggregation and resign from using @DBRef or @DocumentReference (I encountered the same problem using @DocumentReference)? If yes, then @DocumentReference does not have many use cases as I can see that for around 100 entities fetching data works very slow. @DBRef at least can fetch data from another database.

I would greatly appreciate any solution regarding this matter.

Thank you.

**Comment by dominikpiegdon** [comment]
Hey @mp911de !

Thank you for your answer and clarifying this framework behaviour. So as I understand, @DocumentReference and @DBRef use cases are rather limited to pretty small amount of data (or lazy fetched), is that right?

Apart this, I don't have more questions. I guess I can live with aggregations as I don't expect to use them too often.

**Comment by mp911de** [comment]
>  So as I understand, @DocumentReference and @DBRef use cases are rather limited to pretty small amount of data (or lazy fetched), is that right?

Yes, if you fetch 100's of documents, then the latency will severely impact your application  while fetching one or two documents will likely go unnoticed.

Closing this one as the issue is answered.
